torch/distributed/fsdp/_optim_utils.py

```diff
@@ -285,7 +285,16 @@ def _unflatten_communicated_optim_state(
                 views = flat_param_views[state_name]
             optim_state: Union[torch.Tensor, ShardedTensor, DTensor] = next(views)
             if shard_state:
-                if not fsdp_state._optim_state_dict_config.use_dtensor:
+                osd_config = fsdp_state._optim_state_dict_config
+                if (
+                    hasattr(osd_config, "_use_dtensor")
+                    and osd_config._use_dtensor is True
```

Comment:
nit: 
```suggestion
                    getattr(osd_config, "_use_dtensor", False)
```

Reply:
When can `_use_dtensor` not be defined on `osd_config`?

Reply:
> When can `_use_dtensor` not be defined on `osd_config`?

Previously, _use_dtensor is defined in OptimStateDictConfig. I remove the field in OptimStateDictConfig and move it inside ShardedOptimStateDictConfig, since it doesn't really make sense for FullOptimStateDictConfig to have this field.